urdfModifiers/core/fixedOffsetModifier.py

```diff
@@ -0,0 +1,147 @@
+            significant_length *= 2
+        return significant_length
+
+    def get_joint_origin(self, joint):
+        """Returns the origin of a joint"""
+        return (matrix_to_xyz_rpy(joint.origin) if joint else None)
+
+    def get_link_origin(self, link):
+        """Returns the origin of a joint"""
+        return matrix_to_xyz_rpy(link.visuals[0].origin)
+
+    def calculate_offsets(self):
+        link_length = self.get_significant_length()
+        link_origin = self.get_link_origin(self.link)
+        link_origin_z = link_origin[2]
+        child_joint_origin = self.get_joint_origin(self.child_joint)
+        child_joint_origin_z = (child_joint_origin[2] if child_joint_origin is not None else 0)
+        parent_joint_offset = (link_origin_z - link_length * math.copysign(1, child_joint_origin_z) / 2 if self.parent_joint else None)
+        child_joint_offset = (link_origin_z + link_length * math.copysign(1, child_joint_origin_z) / 2 - child_joint_origin_z if self.child_joint else None)
+        return parent_joint_offset, child_joint_offset
+
+    def modify(self, modifications):
+        """Performs the modifications in the link-joint setup"""
+
+        trivial_modifications = Modification()
+        if modifications.radius:
+            trivial_modifications.add_radius(modifications.radius.value, modifications.radius.absolute)
+        if modifications.density:
+            trivial_modifications.add_density(modifications.density.value, modifications.density.absolute)
+        if modifications.mass:
+            trivial_modifications.add_mass(modifications.mass.value, modifications.mass.absolute)
+        
+        self.link_modifier.modify(trivial_modifications)
+
+        if modifications.dimension:
+            original_length = self.get_significant_length()
+            if modifications.dimension.absolute:
+                self.change_dimension_and_keep_offsets(modifications.dimension.value)
+            else:
+                self.change_dimension_and_keep_offsets(original_length * modifications.dimension.value)
+
+    def change_dimension_and_keep_offsets(self, new_length):
+        """Changes the dimension of the link while keeping the offset between it and both parent and child joints"""
+
+        child_joint_origin = self.get_joint_origin(self.child_joint)
+        child_joint_origin_z = (child_joint_origin[2] if child_joint_origin is not None else 0)
+
+        parent_joint_offset, child_joint_offset = self.calculate_offsets()
+
+        link_modification = Modification()
+
+        if parent_joint_offset is not None:
+            new_link_origin = parent_joint_offset + new_length * math.copysign(1, child_joint_origin_z) / 2
+            link_modification.add_position(new_link_origin, absolute=True)
+        else:
+            # for the joint calculations, if there is no parent we position it as if it were in the center of the visual
+            parent_joint_offset = -new_length / 2
```

Comment:
How did you derive this equation?

Reply:
This formula assumes that the parent (actually non-existent) joint origin is directly at the "start" of a link. This in turn allows the calculation of the modified child joint origins to work smoothly with the other formulas.

you can think of it as s_o = v_o - v_l / 2 * sign(j_o), but where v_o is 0 since there is no parent joint

Reply:
I see, that make sense. But in this case, shouldn't there be a `sign(j_o)` or similar term?

Reply:
Correct, I will add it. We assumed, being the parent link, that the sign would always be positive in this case, but it is more general/makes more sense/is more reliable if we include the sign here.